### ISSUE TITLE
Integration with scalafix for automated code fixes

### ISSUE THREAD

**Issue by Daenyth** [fthomas/scala-steward#149]
Is there a way to integrate scala-steward with scalafix, so that when a library maintainer supplied a scalafix migration from one version to another (such as the one in the works for fs2 0.10 to 1.0), that steward can automatically apply it in the PR along with the build change?

**Comment by fthomas** [comment]
That's a great idea! How would one invoke the scalafix migration for fs2? There is nothing yet in scala-steward to do this but let's figure out what would be needed to do it.

**Comment by Daenyth** [comment]
I'm not sure what's the best way to do it - I think for MVP, perhaps a config file on the steward side can list an index of `(artifactId, fromVersion, toVersion, scalaFixDependencyArtifactId, scalafixInvocationCommand)`. The dependency being for cases like the http4s 0.20 upgrade, where you will add a dependency to build.sbt in order to run the fix. It could also be used for the git address in the case of github-style fixes.

Once the infrastructure is in place to run it at all, the manual config driven approach could be revisited and an automatic method could be written.

**Comment by fthomas** [comment]
Sounds like a plan. It seems the github-style fixes are easier to run since they don't require changes to the `build.sbt`. At least this is the impression I got from reading https://github.com/typelevel/cats/blob/master/scalafix/README.md which also suggests that github-style fixes only require that sbt-scalafix is added to the build.

**Comment by Daenyth** [comment]
Github-style fixes seem like a good MVP target

**Comment by olafurpg** [comment]
Awesome idea! For now, it's probably easiest to start with `github:org/repo` style rewrites but once https://github.com/scalacenter/scalafix/issues/905 is fixed it will be possible to run published scalafix linter rules with one-liners from the shell.

**Comment by amarrella** [comment]
Awesome idea! I'll refactor the fs2 scalafix in the works to be github compatible for now so we can test it as well :)

**Comment by amarrella** [comment]
Ok you can already test the fs2 rule here: 
```
sbt ";scalafixEnable;scalafix github:amarrella/fs2/v1?sha=672ea4f9" 
```

**Comment by amarrella** [comment]
Maybe we could fork and try on this repo? https://github.com/pauljamescleary/scala-pet-store

**Comment by fthomas** [comment]
Thanks, @amarrella. Let's make this a little bit more concrete. Say we make a list of updates and their corresponding rewrite rules using this data type:
```scala
case class Migration(
  groupId: String,
  artifactId: String,
  fromVersion: String,
  toVersion: String,
  gitHubRewrite: String
)
```
Would these entries make sense:
```scala
Migration("co.fs2", "fs2-core", "0.10.6",   "1.0.0", "github:amarrella/fs2/v1?sha=672ea4f9")
Migration("co.fs2", "fs2-core", "1.0.0-M1", "1.0.0", "github:amarrella/fs2/v1?sha=672ea4f9")
```
? Can the same `gitHubRewrite` be used for migrations from 0.10.6 and 1.0.0-M1 to 1.0.0? Maybe the `fromVersion` is superfluous because `github:amarrella/fs2/v1?sha=672ea4f9` can be used for all updates to 1.0.0 independent of the current version?

**Comment by fthomas** [comment]
Forking scala-pet-store sounds good to me. It also seems to be using fs2 0.10 via doobie 0.5. Does your fs2 rewrite changes anything there?

**Comment by amarrella** [comment]
> ? Can the same gitHubRewrite be used for migrations from 0.10.6 and 1.0.0-M1 to 1.0.0? Maybe the fromVersion is superfluous because github:amarrella/fs2/v1?sha=672ea4f9 can be used for all updates to 1.0.0 independent of the current version?

yeah technically the rule doesn't limit where it's ran. It would be probably weird/have incomplete results to run it on a version before 0.10 though. Not sure what's best here.

> Forking scala-pet-store sounds good to me. It also seems to be using fs2 0.10 via doobie 0.5. Does your fs2 rewrite changes anything there?

It just changes the StreamApp to an IOApp

**Comment by fthomas** [comment]
Ok, I'd drop `fromVersion` then and worry about it later. :-) I agree with @Daenyth that we should build a MVP and then refine it later.

**Comment by amarrella** [comment]
Cool! Really excited about this!

**Comment by fthomas** [comment]
Me too! :-) I just forked https://github.com/fthomas/scala-pet-store for devlopment and testing and invited you as collaborators.

**Comment by amarrella** [comment]
Http4s is available for testing too (changes some stuff in the test repo as well)
```
sbt ";scalafixEnable; scalafix github:amarrella/http4s/v0_20?sha=85ef964"
```

**Comment by fthomas** [comment]
Wow, I just ran `sbt ";scalafixEnable;scalafix github:amarrella/fs2/v1?sha=672ea4f9" ` in scala-pet-store. This stuff is too easy. ;-)

So on the scala-steward side we need to
 * maintain a list of migrations
 * for each update check if an applicable migration exists

and if a migration exists:
 * add sbt-scalafix to the project or let scala-steward add sbt-scalafix as global plugin
 * run `sbt ";scalafixEnable;scalafix $migration" ` before creating the pull request

This is straightforward.

**Comment by amarrella** [comment]
> Wow, I just ran sbt ";scalafixEnable;scalafix github:amarrella/fs2/v1?sha=672ea4f9" in scala-pet-store. This stuff is too easy. ;-)

Yeah it's amazing. Kudos to @olafurpg 

> add sbt-scalafix to the project or let scala-steward add sbt-scalafix as global plugin

My vote goes to have it as global plugin in scala steward.

If we want to be really fancy, we should also run scalafmt in the projects that have it installed (scalafixes mess formatting). One issue with this might be if the scalafixed code doesn't compile - which is quite common for now, not sure if scalafmt works in this case.

**Comment by amarrella** [comment]
What about multiple scalafixes? Would they be separate prs or one pr? The issue i see with one pr only might be if the result of one doesn't compile, all successive scalafixes won't even start. Can we do separate prs or would it change how scala-steward works too much?

Alternatively I "think" we could chain them in one `sbt ";scalafixEnable;scalafix $migration1;scalafix $migration2;" ` but i'm not entirely sure it will work

**Comment by amarrella** [comment]
> Alternatively I "think" we could chain them in one sbt ";scalafixEnable;scalafix $migration1;scalafix $migration2;" but i'm not entirely sure it will work

Confirmed, this seems to do both in one go skipping the compilation problems
```
sbt ";scalafixEnable;scalafix github:amarrella/fs2/v1?sha=672ea4f9 github:amarrella/http4s/v0_20?sha=85ef964"
```

I think then migration should be called migrations and be a list to support this case

**Comment by fthomas** [comment]
@amarrella scala-steward creates one PR per update. An update changes the version of a single artifact or of a group of artifacts with the same groupId and versions. So it will open one PR for fs2 and one for http4s. If two or more dependencies need to be updated in lockstep, a human needs to do it.

**Comment by amarrella** [comment]
👍 cool, things are simpler then. 

Another thing that we need to pay attention too is memory usage / metaspace particularly as scalafix "compiles" the code it's going to fix to a SemanticDB file.

**Comment by amarrella** [comment]
This seems good for testing too: https://github.com/kubukoz/brick-store

**Comment by olafurpg** [comment]
Any suggestion for the one-line syntax for running published scalafix rules? My current idea is

```
> scalafix dependency:SemanticRule@com.geirsson:example-scalafix-rule:1.3.0
```

**Comment by fthomas** [comment]
Thanks for the update, @olafurpg. The current WIP implementation (#153) should work whether the rewrite is a GitHub source rewrite or a published rewrite since they only differ in the argument that is passed to the `scalafix` sbt command if I understand correctly.

The only thing missing until we can merge #153 is reformatting the scalafix changes with scalafmt (or something similar). The current scalafix changes are not really human-readable as can be seen for example here: https://github.com/fthomas/scala-pet-store/pull/13

**Comment by olafurpg** [comment]
It looks like the http4s rewrite is using `replaceTree` on large tree nodes which is not recommended since it removes comments and other formatting details. Scalafix rules should ideally operate at the token level to produce minimal diffs and preserve comments.

**Comment by fthomas** [comment]
Nice, then there should be no need for scala-steward to run scalafmt after the scalafix rewrites. @amarrella could you tweak the fs2 rewrite such that it produces a minimal diff?

**Comment by amarrella** [comment]
@olafurpg thanks for the tip! :) Can you give me a few examples on how they would look like for my fix?


@fthomas I'll try to do it sometime next week, though i'm super busy lately! :( I accept prs though :D (I say I, as the fix has not been merged to fs2 yet)

**Comment by olafurpg** [comment]
@amarrella A rely should use `replaceTree` sparingly and only for small tree nodes that are unlikely to have comments. Other rewrites should ideally use `addRight`/`addLeft`/`removeTokens` https://scalacenter.github.io/scalafix/docs/developers/patch.html

**Comment by amarrella** [comment]
Apologies for the delay, it has been a very busy period for me :(
For now I improved the rules wrt StreamApp/IOApp, now change is smaller and should keep formatting / comments
`sbt ";scalafixEnable;scalafix github:amarrella/fs2/v1?sha=scalafix"` to try

**Comment by fthomas** [comment]
I've just merged #153 which means that scala-steward now applies applicable rules listed in [`migrations`](https://github.com/fthomas/scala-steward/blob/master/modules/core/src/main/scala/org/scalasteward/core/scalafix/package.scala). One example PR with a rule applied is: https://github.com/fthomas/scalafix-test/pull/4/files

The next step would be to collect more migration rules and try them out in the wild to find potential bugs or maybe identify missing metadata in [`Migration`](https://github.com/fthomas/scala-steward/blob/master/modules/core/src/main/scala/org/scalasteward/core/scalafix/Migration.scala).

Once this feature is stable and we're confident that nothing is missing in `Migration` we should figure out a common mechanism for libraries to publish the metadata in [`migrations`](https://github.com/fthomas/scala-steward/blob/master/modules/core/src/main/scala/org/scalasteward/core/scalafix/package.scala) themselves such that scala-steward can discover it automatically without the need to change its sources.

**Comment by amarrella** [comment]
@fthomas this is fantastic!

**Comment by fthomas** [comment]
I'm closing this because the feature has been implemented in #153. For further improvements, I've opened these follow-up tickets:
 * https://github.com/fthomas/scala-steward/issues/593: Show applied migrations in the PR body
 * https://github.com/fthomas/scala-steward/issues/603: Document the Scalafix integration
 * https://github.com/fthomas/scala-steward/issues/604: Consider using a separate commit for Scalafix changes
